lib/services/consumed/autocomplete.js

```diff
@@ -0,0 +1,130 @@
+/* @flow */
+import { CompositeDisposable, Disposable } from "atom";
+
+import type { Store } from "../../store";
+import type WatchesStore from "../../store/watches";
+import type WatchStore from "../../store/watch";
+
+/**
+ * This acts as a global storage for the consumed service.
+ */
+export class AutocompleteWatchEditor {
+  disposables = new CompositeDisposable();
+  /** The `consumed autocompleteWatchEditor` */
+  addAutocompleteToEditor = (
+    editor: atom$TextEditor,
+    labels: Array<string>
+  ) => {
+    return;
+  };
+
+  /**
+   * This function is called on activation of autocomplete, or if autocomplete is
+   * already active, then it is called when hydrogen activates.
+   *
+   * @param {Store} store - The global Hydrogen store.
+   * @param {Function} watchEditor - The function provided by `autocomplete.watchEditor`.
+   * @returns {Disposable} This is for clean up when autocomplete or hydrogen deactivate.
+   */
+  consume(store: Store, watchEditor: atom$AutocompleteWatchEditor) {
+    this.addAutocompleteToEditor = watchEditor;
+
+    // Add autocomplete capabilities to already existing watches
+    for (let kernel of store.runningKernels) {
+      for (let watch of kernel.watchesStore.watches) {
+        this.addAutocompleteToWatch(kernel.watchesStore, watch);
+      }
+      this.disposables.add(kernel.watchesStore.autocompleteDisposables);
+    }
+
+    const disposable = new Disposable(() => this.disable(store));
+    store.subscriptions.add(disposable);
+    return disposable;
+  }
+
+  /**
+   * This function is just for cleaning up when either autocomplete or hydrogen is deactivating.
+   *
+   * @param {Store} store - The global Hydrogen store.
+   */
+  disable(store: Store) {
+    // Removes the consumed function `watchEditor`
+    this.addAutocompleteToEditor = (
+      editor: atom$TextEditor,
+      labels: Array<string>
+    ) => {
+      return;
+    };
+
+    /*
+     * Removes disposables from all watches (leaves references inside
+     * `this.disposables` to be disposed at the end).
+     * Autocomplete is only actually disabled on dispose of `this.disposables`
+     */
+    for (let kernel of store.runningKernels) {
+      for (let watch of kernel.watchesStore.watches) {
+        watch.autocompleteDisposable = null;
+      }
+      kernel.watchesStore.autocompleteDisposables = new CompositeDisposable();
```

Comment:
I think it would be clearer if you move these `new CompositeDisposable` stuff into `consume`, since these kind of actions are not `disabling`.